### PR TITLE
Support for #include #insert and validation for closed template

### LINKED ISSUE #438: Support for #include #insert
#include #insert works only for navigation

### PR DESCRIPTION
**Pull request by angelozerr**
Support for #include #insert

Fixes #438

### CODE REVIEW

**Comment by angelozerr** [comment]
the 
QuteValidationWithClosedTemplate

If you open a Qute template (just to start the Qute LS) ,and you close it (or not), and you wait for several  seconds, you should see all Qute errors of the project in the PROBLEMS view: 

image

**Comment by angelozerr** [comment]
@mkouba @FroMage I'm supporting completion for template ids in #include. Here a screenshot with the https://github.com/FroMage/quarkus-renarde-todo/blob/7784cba6de7f31e937726eec6725df3c95eca1f6/src/main/resources/templates/Login/login.html#L1

image

As you can see, the completion shows the `.html`, `.txt` extensions, but in doc https://quarkus.io/guides/qute-reference#include_helper sample doesn't use the `.html` file extension. So I assume that we can write:

`{#include base}` or `{#include base.html}`

My question is what is the prefered syntax (with or without html?) Completion should show:

 * 1) template id with file extension?
 * 2) template id without file extension?
 * 3) template id with and without file extension both?

Should we have a settings to decide if 1), 2), or 3) should be done?

**Comment by FroMage** [comment]
Well, good question. I didn't know you could do `{#include base}` without extension. Does it find the correct extension in case of several alternatives?

- `{#include email}` in `Emails/hello.txt` will include `email.txt`
- `{#include email}` in `Emails/hello.html` will include `email.html`

I guess that's a question for @mkouba I never tried that. I should try it.

As for the completion, I'd start by suggesting that you put toplevel templates first, which will get you `main.html` (with or without extension, depending on the outcode of this discussion) first, because in my experience, includes are 100% always including base templates at the root, never anything else.

**Comment by mkouba** [comment]
> Should we have a settings to decide if 1), 2), or 3) should be done?

That's a good question and I don't have a simple answer ;-).

> Well, good question. I didn't know you could do `{#include base}` without extension. Does it find the correct extension in case of several alternatives?
> 
>     * `{#include email}` in `Emails/hello.txt` will include `email.txt`
> 
>     * `{#include email}` in `Emails/hello.html` will include `email.html`
> 
> 
> I guess that's a question for @mkouba I never tried that. I should try it.

Currently, it does not reflect the selected variant/suffix of the including template but iterates over the suffixes defined in `quarkus.qute.suffixes`, i.e. the default value is `qute.html,qute.txt,html,txt` so the `{#include email}` will try `email`, `email.qute.html`, `email.qute.txt`, `email.html` and finally `email.txt`.

In theory, we could reflect this kind of information but if you think about it, when you're inside the `Emails/hello.html` then you know that you're going to include the `html` variant, right? That said, the variant without suffix is IMO easier to read. 

It might make sense to provide the file with suffix if multiple variants exists...

**Comment by angelozerr** [comment]
> As for the completion, I'd start by suggesting that you put toplevel templates first,

Very good feedback! Done:

image

**Comment by angelozerr** [comment]
Thanks @mkouba for your feedback.

If I understand your idea, completion should show template id without suffix, and for the case of conflicts like `email.txt`, `email.html`, completion should show file extension.

@FroMage do you like this idea?

**Comment by FroMage** [comment]
That's a good idea, I think.

**Comment by angelozerr** [comment]
> That's a good idea, I think.

Here the result:

image

**Comment by FroMage** [comment]
Looks great!

**Comment by fbricon** [comment]
looks like completion proposes the current template (Application/index), which sounds like a bad idea

**Comment by angelozerr** [comment]
> Looks great!

Glad the feature please you.

>looks like completion proposes the current template (Application/index), which sounds like a bad idea

Good catch! Fixed:

image

**Comment by datho7561** [comment]
The validation for unopened templates doesn't appear until you open the first Qute template. Is this intentional? I think it could be interesting to validate all the templates as soon as the language server is able to.

**Review by JessicaJHee — COMMENTED** [review]
I've played around with deleting and restoring the templates and validation is working as expected! I wonder if it's a good idea to show the elements we could insert in the completion first? (i.e., have `title` show up before the suggested sections)
Screenshot from 2023-03-14 12-08-09

I also was wondering if we support the simple example in https://quarkus.io/guides/qute-reference#include_helper. I'm not quite sure how it works.

**Comment by datho7561** [comment]
> It should work, what is your problem?

Maybe this is a different issue, but what I'm seeing is if I have the template `"other-template.html"`:

```qute-html
<h3>{#insert myTitle}MY TITLE :D{/}</h3>

<p>
{#insert}
    this is the paragraph used in this template
{/insert}
</p>
```

Then that means I can pass in a `myTitle` when I `{#include}` the template that will replace the `"MY TITLE :D"` value:

```qute-html
{#include other-template.html myTitle="New Title"}
  Content
{/}
```

However, right now we don't have completion or validation for `myTitle` when we include `"other-template.html"`. Is this expected? If so, then I think it's a good idea to open issues for that.

**Comment by angelozerr** [comment]
> However, right now we don't have completion or validation for myTitle when we include "other-template.html". Is this expected? If so, then I think it's a good idea to open issues for that.

Indeed parameter in include is not supported, please create a new issue for that.

**Review by datho7561 — CHANGES_REQUESTED** [review]

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/project/WatchDir.java [hunk 1]
```diff
@@ -0,0 +1,199 @@
+package com.redhat.qute.project;
+
+import static java.nio.file.LinkOption.NOFOLLOW_LINKS;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_CREATE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_DELETE;
+import static java.nio.file.StandardWatchEventKinds.OVERFLOW;
+
+import java.io.IOException;
+/*
+ * Copyright (c) 2008, 2010, Oracle and/or its affiliates. All rights reserved.
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/project/WatchDir.java:2]: fixed

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/settings/capabilities/QuteCapabilityManager.java [hunk 2]
```diff
@@ -126,7 +126,9 @@ public void initializeCapabilities() {
 	private void registerWatchedFiles() {
 		List<FileSystemWatcher> watchers = new ArrayList<>(2);
 		watchers.add(new FileSystemWatcher(Either.forLeft("**/*.html")));
-		watchers.add(new FileSystemWatcher(Either.forLeft("**/*.qute.html")));
+		watchers.add(new FileSystemWatcher(Either.forLeft("**/*.json")));
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/settings/capabilities/QuteCapabilityManager.java:129]: fixed

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/project/QuteTextDocument.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
+/*******************************************************************************
+* Copyright (c) 2021 Red Hat Inc. and others.
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/project/QuteTextDocument.java]: fixed

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/project/QuteProjectFilesWatcher.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+ * src/main/resources/templates folder of a Qute project and generates
+ * the LSP {@link DidChangeWatchedFilesParams} to update closed document cache.
+ * 
+ * This class is used only when the LSP client cannot support
+ * {@link DidChangeWatchedFilesParams}.
+ * 
+ * @author Angelo ZERR
+ *
+ */
+public class QuteProjectFilesWatcher extends WatchDir {
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/project/QuteProjectFilesWatcher.java:38]: I must write a test with this support.

To test in vscode:

 * comment this line https://github.com/redhat-developer/quarkus-ls/pull/810/files#diff-854c916bc4ae147f987affa94adc4054d62afb789c45d2778a4d2b18164af4dfR134
 * comment this line https://github.com/redhat-developer/quarkus-ls/pull/810/files#diff-16409fc750922b54b0bd633df892f47718829028062ea36eb309e704b75b0ac2L131

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/project/QuteProjectRegistry.java [hunk 5]
```diff
@@ -1375,4 +1405,76 @@ public CompletableFuture<String> getJavadoc(QuteJavadocParams params) {
...
+		return projects.values();
+	}
+
+	public void didChangeWatchedFiles(DidChangeWatchedFilesParams params) {
+		Set<QuteProject> projects = new HashSet<>();
+		List<FileEvent> changes = params.getChanges();
+		// Some qute templates are deleted, created, or changed
+		// Collect impacted Qute projects
+		for (FileEvent fileEvent : changes) {
+			if (fileEvent.getType() != FileChangeType.Changed) {
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/project/QuteProjectRegistry.java]: I did a new push and now FileEventType#Changed should work.

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/completions/QuteCompletionForTemplateIds.java [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+			List<QuteTextDocument> documents = new ArrayList<>(project.getDocuments());
+			// Sort template ids to show at first the root files of the
+			// src/main/resources/templates folder (ex : base,main).
+			Collections.sort(documents, TEMPLATE_ID_COMPARATOR);
+
+			// To keep the sort in LSP we generates a basic sort text by adding the 'a'
+			// letter.
+			StringBuilder sortText = new StringBuilder();
+
+			Map<String /* template if with short syntax (ex : base) */, List<QuteTextDocument> /*
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/completions/QuteCompletionForTemplateIds.java]: fixed

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/project/QuteProject.java [hunk 7]
```diff
@@ -65,36 +57,74 @@
  */
 public class QuteProject {
 
+	private static String[] TEMPLATE_VARIANTS = { "", ".html", ".qute.html", ".json", ".qute.json", ".txt", ".qute.txt",
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/project/QuteProject.java]: fixed

**Comment by FroMage** [comment]
> Indeed parameter in include is not supported, please create a new issue for that.

That's now how we pass parameters to included files, though.

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/project/QuteProjectRegistry.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (datho7561) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/project/QuteProjectRegistry.java]: If I fix an error in a file while using a different file, then the error marker goes away in the language client, but if I add an error to the file, then no errors are added to the file.

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/project/QuteProjectFilesWatcher.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/project/QuteProjectFilesWatcher.java:39]: @datho7561 I have written tests with / without  this LSP capability https://github.com/redhat-developer/quarkus-ls/pull/810/files#diff-45a8adb3bad6abdda88d002f45a2b22aa7574e9a08f4e10fc6fef91cfbecb95b (CI build is working).

I retried in vscode by updating this method https://github.com/redhat-developer/quarkus-ls/blob/c0fc3388a170e0f709df88a625ec85d3c89460cd/qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/settings/capabilities/ClientCapabilitiesWrapper.java#L93 (by forcing to return to false) and it works correctly.

Could you retry it please and if it doesn't work, could you give me your usecase. Thanks.

**Comment by angelozerr** [comment]
> That's now how we pass parameters to included files, though.

Thanks for your feedback. We have created the issue for that https://github.com/redhat-developer/quarkus-ls/issues/815

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/project/QuteProjectFilesWatcher.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (datho7561) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/project/QuteProjectFilesWatcher.java:39]: I don't seem to get any responses from the server (other than unmatched cancellation responses) when I force this method to return true.

[the remark this answers is not in the document]
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/project/QuteProjectFilesWatcher.java:39]: You need to force to **false**. vscode support it.

[the remark this answers is not in the document]
Comment (datho7561) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/project/QuteProjectFilesWatcher.java:39]: Does it work for you?

**Review by datho7561 — APPROVED** [review]
Looks good to me, and it's working well. Thanks, Angelo!

**Pull request merged**